qcp/matrix.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2022 Tiernan8r
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+from abc import ABC
+from typing import Union
+from ._types import SCALARS, VECTOR, MATRIX
+
+class Matrix(ABC):
```

Comment:
When a class is defined as an `ABC` (Abstract Base Class), the method definitions should just be method stubs, with any classes inherinting from `Matrix` being where the actual code is implemented, see https://github.com/Tiernan8r/quantum_computing_project/pull/26/files#diff-bf5ed9ed646bcabe407cb419b345dc3b1c87058e6b63b8f338f7bd210276a88bR1 for example.

Ideally all your implementation code here should be moved to a second file, where the Matrix class is defined as `MyMatrix(Matrix)`.